### PR TITLE
feat(web): read an .xlsx workbook into capped per-sheet grids

### PR DESCRIPTION
**Pull request by Jasonnnz**
## Summary

- Adds `parseWorkbook(blob)`, which reads an OOXML workbook into one capped `ParsedCsv` grid per visible sheet: shared, inline, and rich-run strings, booleans, errors, formulas (cached value, else the formula text), and date-styled serials rendered as `yyyy-mm-dd` with `date1904` and the 1900 phantom leap day honoured.
- Sheet and shared-string parts are streamed and abandoned once the row cap is met, so the reader never inflates more of a sheet than the preview shows. A test proves it by leaving broken XML past the cap.
- Exports `looksLikeHeader` from `csv.ts` so a sheet and a CSV pick their header row the same way, and adds `xlsx.test-helper.ts`, which builds workbook containers in memory so no binary fixture enters the repo.
- No production consumer yet: PR 3 renders the grids, PR 4 routes `.xlsx` and `.xlsm` to them.

Part of LUM-3387.
Part of plan: lum-3387-xlsx-drawer-preview.md (PR 2 of 4)

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: clients/web/src/domains/chat/components/local-file/preview/xlsx.ts [hunk 1]
```diff
@@ -0,0 +1,494 @@
...
+    let buffer = "";
+    let searchFrom = 0;
+    let seen = 0;
+    let settled = false;
+
+    stream.on("data", (chunk) => {
+      if (settled) {
+        return;
+      }
+      buffer += chunk;
```
[the remark this answers is not in the document]
Comment (Jasonnnz) [reply clients/web/src/domains/chat/components/local-file/preview/xlsx.ts:200]: Added an exported MAX_PART_CHARS cap (64 MiB of characters) alongside the row cap: readBoundedPart now cuts a part at the start of the last complete marker once the buffer passes the cap, so one enormous cell or shared string can no longer inflate a whole part before DOMParser sees it. A part cut before its first marker reads as empty, and a sheet that loses a shared string to the cut reports truncated so the footer says so. parseWorkbook takes an optional { maxPartChars } so tests can exercise it with a small cap.

File: clients/web/src/domains/chat/components/local-file/preview/xlsx.ts [hunk 2]
```diff
@@ -0,0 +1,494 @@
...
+function readSheetRows(
+  doc: Document,
+  isDateStyle: boolean[],
+  date1904: boolean,
+): SheetRows {
+  const rows: RawCell[][] = [];
+  let truncated = false;
+  let highestSharedIndex = -1;
+
+  for (const row of Array.from(doc.getElementsByTagName("row"))) {
```
[the remark this answers is not in the document]
Comment (Jasonnnz) [reply clients/web/src/domains/chat/components/local-file/preview/xlsx.ts]: readSheetRows now honours the row's r attribute: a gap fills with empty rows before the present row is appended, counted against MAX_CSV_ROWS, so rows 1 and 3 render as 1 and 3. Filling stops at the cap and marks the grid truncated, no trailing blanks are added past the last present row, and a row with no r still appends after the previous one.

File: clients/web/src/domains/chat/components/local-file/preview/xlsx.ts [hunk 3]
```diff
@@ -0,0 +1,494 @@
...
+  }
+
+  const value = cachedValue(cell);
+  if (value === null) {
+    // openpyxl writes formulas with no cached result. Showing the formula
+    // beats showing a blank where the user knows there is data.
+    const formula = cell.getElementsByTagName("f")[0];
+    return formula === undefined ? "" : `=${formula.textContent ?? ""}`;
+  }
+  const asNumber = Number(value);
```
[the remark this answers is not in the document]
Comment (Jasonnnz) [reply clients/web/src/domains/chat/components/local-file/preview/xlsx.ts:471]: readCell now treats an empty cached value the same as a missing one, so an unevaluated formula written as <f>...</f><v/> renders as =SUM(A1:A2) rather than 0. A cached <v>3</v> still wins, and an empty <v/> with no formula stays blank.

File: clients/web/src/domains/chat/components/local-file/preview/xlsx.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (Jasonnnz) [reply clients/web/src/domains/chat/components/local-file/preview/xlsx.ts:200]: Addressed in direct push to run-plan/lum-3387-xlsx-drawer/pr-2

File: clients/web/src/domains/chat/components/local-file/preview/xlsx.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (Jasonnnz) [reply clients/web/src/domains/chat/components/local-file/preview/xlsx.ts]: Addressed in direct push to run-plan/lum-3387-xlsx-drawer/pr-2

File: clients/web/src/domains/chat/components/local-file/preview/xlsx.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (Jasonnnz) [reply clients/web/src/domains/chat/components/local-file/preview/xlsx.ts:471]: Addressed in direct push to run-plan/lum-3387-xlsx-drawer/pr-2

**Comment by Jasonnnz** [comment]
@codex review this PR again, the previous issues have been fixed in commit c17b850b01bea589e5bccad402ff64bc34677736

**Comment by Jasonnnz** [comment]
@devin review this PR again, the previous issues have been fixed in commit c17b850b01bea589e5bccad402ff64bc34677736

File: clients/web/src/domains/chat/components/local-file/preview/xlsx.ts [hunk 4]
```diff
@@ -0,0 +1,578 @@
...
+  }
+  return doc;
+}
+
+async function readPart(zip: JSZip, path: string): Promise<string | null> {
+  const entry = zip.file(path);
+  if (entry === null) {
+    return null;
+  }
+  return entry.async("string");
```
[the remark this answers is not in the document]
Comment (Jasonnnz) [reply clients/web/src/domains/chat/components/local-file/preview/xlsx.ts]: Fixed. `xl/workbook.xml`, `xl/_rels/workbook.xml.rels`, and `xl/styles.xml` now read through the same streaming primitive as the sheets, under the same `maxPartChars` cap, so nothing inflates whole any more. A metadata part past the cap throws an error naming the part rather than being truncated, since a cut workbook or style table is worse than no preview; an absent optional part still reads as `null`.

File: clients/web/src/domains/chat/components/local-file/preview/xlsx.ts [hunk 5]
```diff
@@ -0,0 +1,578 @@
...
+    return Number.isInteger(index) && index >= 0 ? { sharedIndex: index } : "";
+  }
+  if (type === "inlineStr") {
+    return joinTextRuns(cell.getElementsByTagName("is")[0]);
+  }
+  if (type === "b") {
+    return cachedValue(cell) === "1" ? "TRUE" : "FALSE";
+  }
+  if (type === "e" || type === "str" || type === "d") {
+    return cachedValue(cell) ?? "";
```
[the remark this answers is not in the document]
Comment (Jasonnnz) [reply clients/web/src/domains/chat/components/local-file/preview/xlsx.ts]: Fixed. The empty-or-missing cached value plus `<f>` check now runs ahead of every typed branch except `s` and `inlineStr`, so an unevaluated `t="str"` or `t="b"` formula renders its formula text instead of `""` or `FALSE`. Cells that carry a cached value keep their typed rendering, and a typed empty cell with no `<f>` is unchanged.

File: clients/web/src/domains/chat/components/local-file/preview/xlsx.ts [hunk 6]
```diff
@@ -0,0 +1,578 @@
...
+
+/**
+ * Render a date serial as `yyyy-mm-dd`, plus ` hh:mm` when the serial carries
+ * a time. The 1900 workbook counts a 29 February 1900 that never existed, so
+ * serials below 60 sit one day behind the real calendar.
+ */
+function formatDateSerial(serial: number, date1904: boolean): string {
+  const epoch = date1904 ? Date.UTC(1904, 0, 1) : Date.UTC(1899, 11, 30);
+  const days = !date1904 && serial < 60 ? serial + 1 : serial;
+  const at = epoch + Math.round(days * MS_PER_DAY);
```
[the remark this answers is not in the document]
Comment (Jasonnnz) [reply clients/web/src/domains/chat/components/local-file/preview/xlsx.ts:392]: Fixed. `formatDateSerial` now special-cases the 1900 system for `60 <= serial < 61` and renders `1900-02-29`, carrying the same ` hh:mm` suffix when the serial holds a time. Serial 59 still renders `1900-02-28` and serial 61 still renders `1900-03-01`.

File: clients/web/src/domains/chat/components/local-file/preview/xlsx.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (Jasonnnz) [reply clients/web/src/domains/chat/components/local-file/preview/xlsx.ts]: Addressed in direct push to run-plan/lum-3387-xlsx-drawer/pr-2

File: clients/web/src/domains/chat/components/local-file/preview/xlsx.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (Jasonnnz) [reply clients/web/src/domains/chat/components/local-file/preview/xlsx.ts]: Addressed in direct push to run-plan/lum-3387-xlsx-drawer/pr-2

File: clients/web/src/domains/chat/components/local-file/preview/xlsx.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (Jasonnnz) [reply clients/web/src/domains/chat/components/local-file/preview/xlsx.ts:392]: Addressed in direct push to run-plan/lum-3387-xlsx-drawer/pr-2

**Comment by Jasonnnz** [comment]
@codex review this PR again, the previous issues have been fixed in commit fffb51d3dfb48bbd6c9baaaaba2e298806618e99

**Comment by Jasonnnz** [comment]
@devin review this PR again, the previous issues have been fixed in commit fffb51d3dfb48bbd6c9baaaaba2e298806618e99

File: clients/web/src/domains/chat/components/local-file/preview/xlsx.ts [hunk 7]
```diff
@@ -0,0 +1,626 @@
...
+    await readPart(zip, "xl/styles.xml", maxPartChars),
+  );
+
+  // A workbook whose sheets are every one hidden still has something to show.
+  const visible = sheets.filter((sheet) => !sheet.hidden);
+  const kept = visible.length > 0 ? visible : sheets;
+
+  const collected: CollectedSheet[] = [];
+  let highestSharedIndex = -1;
+  for (const sheet of kept) {
```
[the remark this answers is not in the document]
Comment (Jasonnnz) [reply clients/web/src/domains/chat/components/local-file/preview/xlsx.ts]: Fixed by making sheet reads lazy rather than adding another cap: `parseWorkbook` now inflates only the bounded metadata parts (`xl/workbook.xml`, its rels, `xl/styles.xml`) and returns sheets whose `read()` inflates that one sheet part on first call and caches the promise, so at most the sheet on screen is materialized and the caps no longer multiply by sheet count. The shared string table is read lazily too, bounded to the highest index the asking sheet needs, cached as `{ strings, readUpTo, exhausted }`, and re-read only when a later sheet reaches further. A sheet whose part is missing or malformed rejects from its own `read()` without affecting the others.

File: clients/web/src/domains/chat/components/local-file/preview/xlsx.ts [hunk 8]
```diff
@@ -0,0 +1,626 @@
...
+        ? undefined
+        : targets.get(sheet.relationshipId);
+    const entry = target === undefined ? null : zip.file(target);
+    if (entry === null) {
+      collected.push({ name: sheet.name, rows: [], truncated: false });
+      continue;
+    }
+    // Sheets are read one at a time so only one part is ever inflated.
+    const part = await readBoundedPart(entry, {
+      marker: { text: "<row", limit: MAX_CSV_ROWS },
```
[the remark this answers is not in the document]
Comment (Jasonnnz) [reply clients/web/src/domains/chat/components/local-file/preview/xlsx.ts]: Fixed in the marker scan: a match now counts only when the character after the marker ends the tag name (whitespace, `>`, or `/`), with a chunk-boundary case that defers the decision when that character has not arrived yet, so `<rowBreaks>` is no longer counted as a row. The same delimiter rule applies to the `<si` marker. Tests cover exactly `MAX_CSV_ROWS` rows plus a `rowBreaks` section parsing with `truncated: false`, and `MAX_CSV_ROWS + 1` rows plus the same section cutting to `truncated: true`.

File: clients/web/src/domains/chat/components/local-file/preview/xlsx.ts [hunk 9]
```diff
@@ -0,0 +1,626 @@
...
+function readWorkbookStructure(xml: string): WorkbookStructure {
+  const doc = parseXml(xml, "xl/workbook.xml");
+  const dateMode = doc
+    .getElementsByTagName("workbookPr")[0]
+    ?.getAttribute("date1904");
+  const sheets = Array.from(doc.getElementsByTagName("sheet")).map((sheet) => {
+    const state = sheet.getAttribute("state");
+    return {
+      name: sheet.getAttribute("name") ?? "",
+      relationshipId: sheet.getAttribute("r:id"),
```
[the remark this answers is not in the document]
Comment (Jasonnnz) [reply clients/web/src/domains/chat/components/local-file/preview/xlsx.ts]: Fixed with two helpers used throughout: `childrenNamed(parent, localName)` matches elements by `localName` regardless of prefix, and `attributeNamed(element, localName, namespaceUri?)` scans `element.attributes`, preferring the namespace match and falling back to a plain local-name match. The sheet relationship id is now read through the relationships namespace instead of the literal `r:id`. Note on the element helper: `getElementsByTagNameNS("*", name)` returns nothing under happy-dom, so the scan filters `getElementsByTagName("*")` on `localName`, which works in both and measured no slower; the reason is in a comment. A new test parses a workbook whose parts prefix every element and the relationship attribute and asserts the same grid as the unprefixed fixture.

File: clients/web/src/domains/chat/components/local-file/preview/xlsx.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (Jasonnnz) [reply clients/web/src/domains/chat/components/local-file/preview/xlsx.ts]: Addressed in direct push to run-plan/lum-3387-xlsx-drawer/pr-2

File: clients/web/src/domains/chat/components/local-file/preview/xlsx.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (Jasonnnz) [reply clients/web/src/domains/chat/components/local-file/preview/xlsx.ts]: Addressed in direct push to run-plan/lum-3387-xlsx-drawer/pr-2

File: clients/web/src/domains/chat/components/local-file/preview/xlsx.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (Jasonnnz) [reply clients/web/src/domains/chat/components/local-file/preview/xlsx.ts]: Addressed in direct push to run-plan/lum-3387-xlsx-drawer/pr-2

**Comment by Jasonnnz** [comment]
@codex review this PR again, the previous issues have been fixed in commit 9dabf3c7613f3f22dfd1b2994ae0288b55d1646a

**Comment by Jasonnnz** [comment]
@devin review this PR again, the previous issues have been fixed in commit 9dabf3c7613f3f22dfd1b2994ae0288b55d1646a
